# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1346342704 14400
#      Thu Aug 30 12:05:04 2012 -0400
# Node ID ad1a980b0cb572507017f22f6b96e9a0cbaa8389
# Parent  6be46886099fa5309b09a0498a6a37e7b74c6299
install default values before initializing interpreter (bug #37161)

* octave.cc (octave_initialize_interpreter): Don't call
install_defaults here.
(octave_main): Call install_defaults here.
* main.cc (main): Likewise.
* main-cli.cc (main): Likewise.
* src/Makefile.am (AM_CPPFLAGS): Include -I$(srcdir)/../libcruft/misc,
-I$(srcdir)/../liboctave, and -I$(top_builddir)/libinterp/interpfcn in
the list.

diff --git a/libinterp/octave.cc b/libinterp/octave.cc
--- a/libinterp/octave.cc
+++ b/libinterp/octave.cc
@@ -41,17 +41,17 @@ along with Octave; see the file COPYING.
 #include "f77-fcn.h"
 #include "file-ops.h"
 #include "file-stat.h"
 #include "lo-error.h"
 #include "oct-env.h"
 #include "pathsearch.h"
 #include "str-vec.h"
 
-#include <defaults.h>
+#include "defaults.h"
 #include "Cell.h"
 #include "defun.h"
 #include "display.h"
 #include "error.h"
 #include "file-io.h"
 #include "input.h"
 #include "lex.h"
 #include "load-path.h"
@@ -682,16 +682,18 @@ maximum_braindamage (void)
 // EMBEDDED is declared int instead of bool because this function is
 // declared extern "C".
 
 int
 octave_main (int argc, char **argv, int embedded)
 {
   octave_process_command_line (argc, argv);
 
+  install_defaults ();
+
   octave_initialize_interpreter (argc, argv, embedded);
 
   return octave_execute_interpreter ();
 }
 
 void
 octave_process_command_line (int argc, char **argv)
 {
@@ -868,21 +870,19 @@ octave_initialize_interpreter (int argc,
   octave_env::set_program_name (argv[0]);
 
   octave_program_invocation_name = octave_env::get_program_invocation_name ();
   octave_program_name = octave_env::get_program_name ();
 
   octave_thread::init ();
 
   // The order of these calls is important.  The call to
-  // install_defaults must come before install_builtins because
-  // default variable values must be available for the variables to be
-  // installed, and the call to install_builtins must come before the
-  // options are processed because some command line options override
-  // defaults by calling bind_internal_variable.
+  // install_builtins must come before the option settings are processed
+  // because some command line options override defaults by calling
+  // bind_internal_variable.
 
   init_signals ();
 
   sysdep_init ();
 
   octave_ieee_init ();
 
   // The idea here is to force xerbla to be referenced so that we will
@@ -892,18 +892,16 @@ octave_initialize_interpreter (int argc,
 
   if (octave_NaN == -1)
     F77_FUNC (xerbla, XERBLA) ("octave", 13 F77_CHAR_ARG_LEN (6));
 
   initialize_error_handlers ();
 
   initialize_default_warning_state ();
 
-  install_defaults ();
-
   initialize_pathsearch ();
 
   if (! embedded)
     install_signal_handlers ();
   else
     quit_allowed = false;
 
   initialize_file_io ();
diff --git a/src/Makefile.am b/src/Makefile.am
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -17,17 +17,20 @@
 # You should have received a copy of the GNU General Public License
 # along with Octave; see the file COPYING.  If not, see
 # <http://www.gnu.org/licenses/>.
 
 include $(top_srcdir)/build-aux/common.mk
 
 ## Search local directories before those specified by the user.
 AM_CPPFLAGS = \
+  -I$(srcdir)/../libcruft/misc \
+  -I$(srcdir)/../liboctave \
   -I$(srcdir)/../libinterp \
+  -I$(top_builddir)/libinterp/interpfcn \
   -I$(top_builddir)/libgnu -I$(top_srcdir)/libgnu
 
 OCTAVE_BINARIES = \
   octave \
   octave-cli
 
 if AMCOND_BUILD_COMPILED_AUX_PROGRAMS
 bin_PROGRAMS = \
diff --git a/src/main-cli.cc b/src/main-cli.cc
--- a/src/main-cli.cc
+++ b/src/main-cli.cc
@@ -19,19 +19,22 @@ along with Octave; see the file COPYING.
 <http://www.gnu.org/licenses/>.
 
 */
 
 #ifdef HAVE_CONFIG_H
 #include <config.h>
 #endif
 
-#include <octave.h>
+#include "defaults.h"
+#include "octave.h"
 
 int
 main (int argc, char **argv)
 {
   octave_process_command_line (argc, argv);
 
+  install_defaults ();
+
   octave_initialize_interpreter (argc, argv, 0);
 
   return octave_execute_interpreter ();
 }
diff --git a/src/main.cc b/src/main.cc
--- a/src/main.cc
+++ b/src/main.cc
@@ -19,26 +19,29 @@ along with Octave; see the file COPYING.
 <http://www.gnu.org/licenses/>.
 
 */
 
 #ifdef HAVE_CONFIG_H
 #include <config.h>
 #endif
 
-#include <octave.h>
-#include <octave-gui.h>
+#include "defaults.h"
+#include "octave.h"
+#include "octave-gui.h"
 
 int
 main (int argc, char **argv)
 {
   int retval = 0;
 
   octave_process_command_line (argc, argv);
 
+  install_defaults ();
+
   if (octave_starting_gui ())
     retval = octave_start_gui (argc, argv);
   else
     {
       octave_initialize_interpreter (argc, argv, 0);
 
       retval = octave_execute_interpreter ();
     }
